File: java/aircraft-positions/src/main/java/com/thehecklers/aircraftpositions/PositionRetriever.java
package com.thehecklers.aircraftpositions;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.client.WebClient;

@AllArgsConstructor
@Component
public class PositionRetriever {
    private final AircraftRepository repository;
    private final WebClient client;

    Iterable<Aircraft> retrieveAircraftPositions(String endpoint) {
        repository.deleteAll();

        client.get()
                .uri((null != endpoint) ? endpoint : "")
                .retrieve()
                .bodyToFlux(Aircraft.class)
                .filter(ac -> !ac.getReg().isEmpty())
                .toStream()
                .forEach(repository::save);

        return repository.findAll();
    }
}

File: kotlin/aircraft-positions/src/main/kotlin/com/thehecklers/aircraftpositions/PositionRetriever.kt
package com.thehecklers.aircraftpositions

import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToFlux

@Component
class PositionRetriever(private val repository: AircraftRepository) {
    private val client = WebClient.create("http://localhost:7634")

    fun retrieveAircraftPositions(): Iterable<Aircraft> {
        repository.deleteAll()

        client.get()
            .uri("/aircraft")
            .retrieve()
            .bodyToFlux<Aircraft>()
            .filter { !it.reg.isNullOrEmpty() }
            .toStream()
            .forEach { repository.save(it) }

        return repository.findAll()
    }
}
